### PR TITLE
{devel}[GCC/10.3.0,intel-compilers/2021.2.0] Boost v1.76.0

### PR DESCRIPTION
**Pull request by boegel**
(created using `eb --new-pr`)

### CODE REVIEW

**Comment by boegel** [comment]
@boegelbot please test @ generoso

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
node2619.swalot.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz (haswell), Python 3.6.8
See https://gist.github.com/2c8cd83f162256fa693f71944bceaa32 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
node3506.doduo.os - Linux RHEL 8.2, x86_64, AMD EPYC 7552 48-Core Processor (zen2), Python 3.6.8
See https://gist.github.com/5084659d4b47b7a38b092cebc803abe1 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
easybuild2.novalocal - Linux centos linux 8.3.2011, POWER, IBM pSeries (emulated by qemu) (power9le), Python 3.6.8
See https://gist.github.com/5cf91ede1dcc3da6ce949226e4ca684c for a full test report.

**Comment by verdurin** [comment]
Test report by @verdurin
**FAILED**
Build succeeded for 1 out of 3 (2 easyconfigs in total)
easybuild-c7.novalocal - Linux centos linux 7.9.2009, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 3.6.8
See https://gist.github.com/77e6dd6e091c73797d180406f79cdd07 for a full test report.

**Comment by verdurin** [comment]
Test report by @verdurin
**FAILED**
Build succeeded for 2 out of 3 (2 easyconfigs in total)
easybuild-c7.novalocal - Linux centos linux 7.9.2009, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 3.6.8
See https://gist.github.com/8638095d528b3c964ba7760dd4361b0f for a full test report.

**Comment by verdurin** [comment]
Test report by @verdurin
**FAILED**
Build succeeded for 1 out of 2 (2 easyconfigs in total)
nuc.lan - Linux Fedora 33, x86_64, Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz, Python 3.9.4
See https://gist.github.com/1d04358f5eacf7b0492baedf1fc94a28 for a full test report.

**Comment by verdurin** [comment]
Test report by @verdurin
**FAILED**
Build succeeded for 0 out of 1 (2 easyconfigs in total)
easybuild-c7.novalocal - Linux centos linux 7.9.2009, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 3.6.8
See https://gist.github.com/54ccd9edbb1f9f9dc91fd881aedf5f57 for a full test report.

**Comment by boegel** [comment]
@verdurin The failing tests boil down to now having `glibc-static` installed, right?
Are existing `Boost-1.74.0-GCC-10.2.0.eb` easyconfigs working as expected on that same system?

**Comment by verdurin** [comment]
As far as I know they are. Note it fails on two different build nodes - Fedora 33 and CentOS 7.

**Comment by Flamefire** [comment]
We need to address 2 issues:
1) The Boost-cmake config files do not work because we build with tagged and system layout where only threading differs, which is ignored (by the config files) for system and hence it tries to use 2 lib files for the same boost lib: `Target Boost::system already has an imported location`
2) We don't build a single-thread Boost, but multithreaded twice which is pointless

1) can be workarounded by `find $EBROOTBOOST/lib/cmake -name '*variant-shared.cmake' -delete` but that is not nice

The best approach (for all Boost versions) is to either only build system layout or build tagged with `threading=single,multi` and optionally add symlinks for the non-tagged versions to either the single- or multi-threaded versions.

Note that in Boost 1.6x (exact version not known yet) the default for threading has changed to `multi`. So it so up to use which one we want to symlink or do so depending on the version.

**Comment by akesandgren** [comment]
If we build both single and multi then default should probably be single, that will cause the least issues with other codes that themselves use threading and/or OpenMP.

File: easybuild/easyconfigs/b/Boost/Boost-1.76.0-GCC-10.3.0.eb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+toolchainopts = {'pic': True}
+
+source_urls = ['https://boostorg.jfrog.io/artifactory/main/release/%(version)s/source/']
+sources = ['%%(namelower)s_%s.tar.gz' % '_'.join(version.split('.'))]
+checksums = ['7bd7ddceec1a1dfdcbdb3e609b60d01739c38390a5f956385a12f3122049f0ca']
+
+dependencies = [
+    ('bzip2', '1.0.8'),
+    ('zlib', '1.2.11'),
+    ('XZ', '5.2.5'),
```
Comment (akesandgren) [inline easybuild/easyconfigs/b/Boost/Boost-1.76.0-GCC-10.3.0.eb:20]: Should probably add zstd and ICU as deps since Boost looks for them

Comment (akesandgren) [reply easybuild/easyconfigs/b/Boost/Boost-1.76.0-GCC-10.3.0.eb:20]: hmm, zstd apparently only needed for VisualC++ builds

**Comment by Flamefire** [comment]
> If we build both single and multi then default should probably be single, that will cause the least issues with other codes that themselves use threading and/or OpenMP.

The default of Boot is multi and AFAIK this "only" means that it is thread-safe, i.e. uses locking and such. So no I wouldn't choose single as the default

**Comment by verdurin** [comment]
Could someone summarise where we are with this one now?

**Comment by Flamefire** [comment]
@verdurin Basically waiting for https://github.com/easybuilders/easybuild-easyblocks/pull/2456 (and a smaller follow-up I'll do afterwards)

**Comment by smoors** [comment]
Test report by @smoors
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2456
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
node223.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz (broadwell), Python 2.7.5
See https://gist.github.com/394ac454f379a311a63bd78835e34f9d for a full test report.

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2456
**FAILED**
Build succeeded for 3 out of 4 (4 easyconfigs in total)
rocky8-eb - Linux rocky linux 8.4, x86_64, Intel(R) Core(TM) i7-6900K CPU @ 3.20GHz (broadwell), Python 3.6.8
See https://gist.github.com/4f5c1c4cf7ea86dba2f4076d6f10607a for a full test report.

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2456
**FAILED**
Build succeeded for 3 out of 4 (4 easyconfigs in total)
centos8-eb - Linux centos linux 8.3.2011, x86_64, Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz (skylake), Python 3.6.8
See https://gist.github.com/617e328a0255eccc6d56d6a918468ea2 for a full test report.

**Comment by SebastianAchilles** [comment]
I see a similar problem as @verdurin with `Boost-1.76.0-intel-compilers-2021.2.0.eb`. The GCC version work and also the older `Boost-1.74.0-iccifort-2020.4.304.eb` on my system. Is this related to having a certain OS packages installed?

**Comment by boegel** [comment]
@SebastianAchilles Can you highlight the problem you're running into? Do you have `glibc-static` installed in the OS?

**Comment by SebastianAchilles** [comment]
`Boost-1.76.0-intel-compilers-2021.2.0.eb` fails on my `centos8` and my `rocky8` minimal container with:
```
> icpx -x c++ -std=c++11 -O3 -s -static -DNDEBUG builtins.cpp class.cpp command.cpp compile.cpp constants.cpp cwd.cpp debug.cpp debugger.cpp execcmd.cpp execnt.cpp execunix.cpp filesys.cpp filent.cpp fileunix.cpp frames.cpp function.cpp glob.cpp hash.cpp hcache.cpp hdrmacro.cpp headers.cpp jam_strings.cpp jam.cpp jamgram.cpp lists.cpp make.cpp make1.cpp md5.cpp mem.cpp modules.cpp native.cpp object.cpp option.cpp output.cpp parse.cpp pathnt.cpp pathsys.cpp pathunix.cpp regexp.cpp rules.cpp scan.cpp search.cpp startup.cpp subst.cpp sysinfo.cpp timestamp.cpp variable.cpp w32_getreg.cpp modules/order.cpp modules/path.cpp modules/property-set.cpp modules/regex.cpp modules/sequence.cpp modules/set.cpp -o b2
/home/easybuild/.local/easybuild/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld: cannot find -lm
/home/easybuild/.local/easybuild/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld: cannot find -ldl
/home/easybuild/.local/easybuild/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld: cannot find -lc
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
> cp b2 bjam
cp: cannot stat 'b2': No such file or directory

Failed to build B2 build engine
```

I think I have `glibc` installed, but not `glibc-static`:
```
[easybuild@rocky8-eb ~]$ yum list installed | grep glibc
Failed to set locale, defaulting to C.UTF-8
glibc.x86_64                       2.28-151.el8                           @System    
glibc-common.x86_64                2.28-151.el8                           @System    
glibc-devel.x86_64                 2.28-151.el8                           @baseos    
glibc-headers.x86_64               2.28-151.el8                           @baseos    
glibc-minimal-langpack.x86_64      2.28-151.el8                           @System
```

**Comment by boegel** [comment]
You wouldn't see this when `glibc-static` is installed, then the linking would work.

So did Boost 1.76.0 start hardcoding `-static`?!

**Comment by Flamefire** [comment]
Yes the `-static` is hardcoded for intel:
       intel-*)
           CXX_VERSION_OPT=${CXX_VERSION_OPT:---version}
           B2_CXXFLAGS_RELEASE="-O3 -s -static"
           B2_CXXFLAGS_DEBUG="-O0 -g -p -static"
       ;;

And that is new in 1.76. We could likely patch this
https://github.com/boostorg/build/commit/5ede4e8d8d8931e52b227f1409c4581785d5e5f0

**Comment by Flamefire** [comment]
@boegel Simply add a sed on tools/build/src/engine/build.sh to delete `-static`

File: easybuild/easyconfigs/b/Boost/Boost-1.76.0-intel-compilers-2021.2.0.eb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+source_urls = ['https://boostorg.jfrog.io/artifactory/main/release/%(version)s/source/']
+sources = ['%%(namelower)s_%s.tar.gz' % '_'.join(version.split('.'))]
+checksums = ['7bd7ddceec1a1dfdcbdb3e609b60d01739c38390a5f956385a12f3122049f0ca']
+
+dependencies = [
+    ('bzip2', '1.0.8'),
+    ('zlib', '1.2.11'),
+    ('XZ', '5.2.5'),
+]
+
```
Comment (SebastianAchilles) [inline easybuild/easyconfigs/b/Boost/Boost-1.76.0-intel-compilers-2021.2.0.eb:19]:
```suggestion

preconfigopts = 'sed -i "s/-static//g" tools/build/src/engine/build.sh && '

```

**Review by SebastianAchilles — COMMENTED** [review]
The suggestion from @Flamefire works for me.

File: easybuild/easyconfigs/b/Boost/Boost-1.76.0-intel-compilers-2021.2.0.eb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+source_urls = ['https://boostorg.jfrog.io/artifactory/main/release/%(version)s/source/']
+sources = ['%%(namelower)s_%s.tar.gz' % '_'.join(version.split('.'))]
+checksums = ['7bd7ddceec1a1dfdcbdb3e609b60d01739c38390a5f956385a12f3122049f0ca']
+
+dependencies = [
+    ('bzip2', '1.0.8'),
+    ('zlib', '1.2.11'),
+    ('XZ', '5.2.5'),
+]
```
Comment (branfosj) [inline easybuild/easyconfigs/b/Boost/Boost-1.76.0-intel-compilers-2021.2.0.eb:18]:
```suggestion
    ('ICU', '69.1'),
]
```

File: easybuild/easyconfigs/b/Boost/Boost-1.76.0-GCC-10.3.0.eb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+]
```
Comment (branfosj) [inline easybuild/easyconfigs/b/Boost/Boost-1.76.0-GCC-10.3.0.eb:21]:
```suggestion
    ('ICU', '69.1'),
]
```

**Review by branfosj — CHANGES_REQUESTED** [review]
Adding ICU, as per @akesandgren comment at https://github.com/easybuilders/easybuild-easyconfigs/pull/12977/files#r643702476

**Comment by branfosj** [comment]
@boegelbot please test @ generoso

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
bear-pg0211u08b.bear.cluster - Linux RHEL 8.3, x86_64, Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz (cascadelake), Python 3.6.8
See https://gist.github.com/f8816666c0a70821da097a96cf9476ca for a full test report.

File: easybuild/easyconfigs/b/Boost/Boost-1.76.0-GCC-10.3.0.eb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    ('zlib', '1.2.11'),
+    ('XZ', '5.2.5'),
+    ('ICU', '69.1'),
+]
+
+configopts = '--without-libraries=python,mpi'
+
+# disable MPI, enable threading support
+boost_mpi = False
+boost_multi_thread = True
```
Comment (branfosj) [inline easybuild/easyconfigs/b/Boost/Boost-1.76.0-GCC-10.3.0.eb:28]:
```suggestion
# disable MPI, build Boost libraries with tagged layout
boost_mpi = False
tagged_layout = True
```

File: easybuild/easyconfigs/b/Boost/Boost-1.76.0-intel-compilers-2021.2.0.eb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    ('ICU', '69.1'),
+]
+
+preconfigopts = 'sed -i "s/-static//g" tools/build/src/engine/build.sh && '
+
+configopts = '--without-libraries=python,mpi'
+
+# disable MPI, enable threading support
+boost_mpi = False
+boost_multi_thread = True
```
Comment (branfosj) [inline easybuild/easyconfigs/b/Boost/Boost-1.76.0-intel-compilers-2021.2.0.eb:27]:
```suggestion
# disable MPI, build Boost libraries with tagged layout
boost_mpi = False
tagged_layout = True
```

**Review by branfosj — CHANGES_REQUESTED** [review]
copy changes due to update to easyblock

File: easybuild/easyconfigs/b/Boost/Boost-1.76.0-GCC-10.3.0.eb (hunk 1 above)
Comment (branfosj) [reply easybuild/easyconfigs/b/Boost/Boost-1.76.0-GCC-10.3.0.eb:20]: done in 8df8360

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
bear-pg0211u08b.bear.cluster - Linux RHEL 8.3, x86_64, Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz (cascadelake), Python 3.6.8
See https://gist.github.com/12b75c0a5cff6bf0c3b55a16ad89c9d7 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
bear-pg0211u16a.bear.cluster - Linux Ubuntu 20.04, x86_64, Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz (cascadelake), Python 3.8.5
See https://gist.github.com/d91850552a0be19589e7c7eb63dcd733 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
bear-pg0306u07a.bear.cluster - Linux RHEL 8.4, POWER, 8335-GTX (power9le), Python 3.6.8
See https://gist.github.com/54c4bfde183a492fd2889cd4e74cf3c9 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
bber0501u03b.bb2.cluster - Linux RHEL 8.3, x86_64, Intel(R) Xeon(R) CPU E5-2690 v3 @ 2.60GHz (haswell), Python 3.6.8
See https://gist.github.com/a80f7cf5e2b4e873d868c63920986f15 for a full test report.

**Comment by Flamefire** [comment]
Test report by @Flamefire
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2513
**FAILED**
Build succeeded for 2 out of 4 (2 easyconfigs in total)
taurusi6607.taurus.hrsk.tu-dresden.de - Linux RHEL 7.9, x86_64, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz (haswell), Python 2.7.5
See https://gist.github.com/77e72a40541dc40a382c0454dc8f68ff for a full test report.

**Comment by Flamefire** [comment]
Test report by @Flamefire
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2513
**FAILED**
Build succeeded for 1 out of 3 (2 easyconfigs in total)
taurusi6607.taurus.hrsk.tu-dresden.de - Linux RHEL 7.9, x86_64, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz (haswell), Python 2.7.5
See https://gist.github.com/b414e626f51389b5765e41854a60e989 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2513
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
bear-pg0211u08b.bear.cluster - Linux RHEL 8.3, x86_64, Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz (cascadelake), Python 3.6.8
See https://gist.github.com/2b2693f4bbb8ff7a2c315e5c450776eb for a full test report.

**Review by branfosj — APPROVED** [review]

**Comment by Flamefire** [comment]
Test report by @Flamefire
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2513
**SUCCESS**
Build succeeded for 3 out of 3 (2 easyconfigs in total)
taurusi6607.taurus.hrsk.tu-dresden.de - Linux RHEL 7.9, x86_64, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz (haswell), Python 2.7.5
See https://gist.github.com/9c3c1485a54c3d48101a02d257b392b4 for a full test report.

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
rocky8-eb - Linux rocky linux 8.4, x86_64, Intel(R) Core(TM) i7-6900K CPU @ 3.20GHz (broadwell), Python 3.6.8
See https://gist.github.com/91df8a3cd38e0dc5efc28e12d0bada0a for a full test report.

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
centos8-eb - Linux centos linux 8.3.2011, x86_64, Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz (skylake), Python 3.6.8
See https://gist.github.com/51c6a733eb7e8056ea3816922e7cc11d for a full test report.

**Comment by smoors** [comment]
Test report by @smoors
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
node361.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz (skylake_avx512), Python 2.7.5
See https://gist.github.com/00af156198fe26c26099b0b45404253c for a full test report.

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
ebvm.novalocal - Linux centos linux 8.3.2011, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 3.6.8
See https://gist.github.com/8f34ce00a6ea40b0dcc1a25590d5a53a for a full test report.

**Review by smoors — APPROVED** [review]
lgtm

**Comment by smoors** [comment]
Going in, thanks @boegel!

**Pull request merged**